c10/core/WrapDimMinimal.h

```diff
@@ -4,37 +4,22 @@
 
 namespace c10 {
 
+namespace detail {
+C10_API int64_t
+maybe_wrap_dim_slow(int64_t dim, int64_t dim_post_expr, bool wrap_scalar);
+}
+
 static inline int64_t maybe_wrap_dim(
     int64_t dim,
     int64_t dim_post_expr,
     bool wrap_scalar = true) {
-  if (dim_post_expr <= 0) {
-    if (!wrap_scalar) {
-      TORCH_CHECK_INDEX(
-          false,
-          "dimension specified as ",
-          dim,
-          " but tensor has no dimensions");
-    }
-    dim_post_expr = 1; // this will make range [-1, 0]
-  }
-
-  int64_t min = -dim_post_expr;
-  int64_t max = dim_post_expr - 1;
-  if (dim < min || dim > max) {
-    TORCH_CHECK_INDEX(
-        false,
-        "Dimension out of range (expected to be in range of [",
-        min,
-        ", ",
-        max,
-        "], but got ",
-        dim,
-        ")");
+  // Inline the fast paths
+  if (C10_LIKELY(-dim_post_expr <= dim && dim < dim_post_expr)) {
```

Comment:
I've checked and an `unsafe_size(dim)` could bring the cost down to `1200` instructions by removing the bounds checking. However, I'm erring on the side of "if you care this much about overhead you should be calling `.sizes()`directly so you can re-use the same ArrayRef".

Reply:
I still like the `unsafe_size(dim)` idea, as it's something that comes up a lot. @ezyang wdyt?